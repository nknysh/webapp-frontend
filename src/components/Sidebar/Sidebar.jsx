import React, { useState } from 'react';
import { map, isEmpty } from 'ramda';
import hash from 'object-hash';

import theme from 'styles/theme';
import { useCurrentWidth } from 'effects';

import { propTypes, defaultProps } from './Sidebar.props';
import { StyledSidebar, SidebarTitle, SidebarLinks, SidebarLink, SidebarChildren, SidebarIcon } from './Sidebar.styles';

export const Sidebar = ({ title, links, isOpen: isOpenProp, children, ...props }) => {
  const currentWidth = useCurrentWidth();
  const [isOpen, setIsOpen] = useState(isOpenProp);

  const onToggle = () => setIsOpen(!isOpen);
  const onClick = () => setIsOpen(false);

  const isMobile = currentWidth <= theme.breakpoints.tablet;
  const hasLinks = !isEmpty(links) || children;

  const renderIcon = () =>
    isMobile && hasLinks && <SidebarIcon>{isOpen ? 'arrow_drop_up' : 'arrow_drop_down'}</SidebarIcon>;

  const renderTitle = () =>
    title && (
      <SidebarTitle onClick={onToggle}>
        {title}
        {renderIcon(isOpen, isMobile)}
      </SidebarTitle>
    );

  const renderLink = link => (
    <SidebarLink key={hash(link)} onClick={onClick}>
      {link}
    </SidebarLink>
  );

  const renderLinks = () => hasLinks && <SidebarLinks isOpen={isOpen}>{map(renderLink, links)}</SidebarLinks>;

  const renderChildren = () =>
    children && (
      <SidebarChildren onClick={onClick} isOpen={isOpen}>
        {children}
      </SidebarChildren>
    );

  return (
    <StyledSidebar {...props}>
      {renderTitle()}
      {renderLinks()}
      {renderChildren()}
    </StyledSidebar>
  );
};

Sidebar.propTypes = propTypes;
Sidebar.defaultProps = defaultProps;

export default Sidebar;
