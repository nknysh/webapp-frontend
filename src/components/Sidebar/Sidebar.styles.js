import styled, { css } from 'styled-components';
import { Icon } from '@material-ui/core';

import theme from 'styles/theme';
import breakpoints from 'styles/breakpoints';
import { Heading1 } from 'styles/typography';

const doubleSpaced = theme.gutter * 2;
const sidebarGutter = `${doubleSpaced}px`;

const sidebarUl = css`
  background: ${theme.colors.white};
  box-shadow: 0 5px 10px ${theme.colors['gray']};
  list-style: none;
  margin: 0;
  transition: ease-in-out 0.25s all;
  position: absolute;
  left: -${sidebarGutter};
  right: -${sidebarGutter};
  padding: ${theme.gutter}px ${sidebarGutter};
  margin-top: -${doubleSpaced}px;

  ${({ isOpen }) =>
    !isOpen &&
    css`
      opacity: 0;
      pointer-events: none;
      padding-top: 0;
    `}

  ${breakpoints.tablet`
    box-shadow: 0 0 0;

    ${({ isOpen }) =>
      !isOpen &&
      css`
        opacity: 1;
        pointer-events: auto;
      `}
  `}
`;

const sidebarLi = css`
  padding: ${doubleSpaced}px 0;
  color: ${theme.primary};
  border-bottom: 1px solid ${theme.colors['gray-medium']};

  &:last-child {
    border-bottom: 0;
  }

  a {
    color: ${theme.primary};
    font-size: ${theme.linkSize}px;
    text-transform: uppercase;
  }
`;

export const StyledSidebar = styled.div`
  position: relative;
`;

export const SidebarTitle = styled(Heading1)`
  border-bottom: 1px solid ${theme.colors['gray']};
  position: relative;
  padding: ${theme.gutter}px ${sidebarGutter} ${doubleSpaced}px 0;
  margin: 0 0 ${doubleSpaced}px;
`;

export const SidebarLinks = styled.ul`
  ${sidebarUl}
`;

export const SidebarLink = styled.li`
  ${sidebarLi}
`;

// For prerendered lists
export const SidebarChildren = styled.div`
  ul {
    ${sidebarUl}

    li {
      ${sidebarLi}
    }
  }
`;

export const SidebarIcon = styled(Icon)`
  position: absolute;
  color: ${theme.colors['black-darker']} !important;
  font-size: 16px !important;
  right: 3px;
  top: ${theme.gutter + 7}px;
  transition: ease-in-out 0.25s all;
`;
