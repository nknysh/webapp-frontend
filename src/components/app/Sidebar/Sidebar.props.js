import PropTypes from 'prop-types';

export const propTypes = {
  title: PropTypes.string,
  links: PropTypes.array,
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.node]),
};

export const defaultProps = {
  links: [],
  isOpen: false,
};
