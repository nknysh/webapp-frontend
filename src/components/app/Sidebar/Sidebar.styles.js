import styled, { css } from 'styled-components';
import { Icon } from '@material-ui/core';

import { theme, breakpoints, Heading1 } from 'styles';

const doubleSpaced = theme.gutter * 2;
const sidebarGutter = `${doubleSpaced}px`;

const sidebarUl = css`
  background: ${theme.colors.white};
  box-shadow: ${theme.boxShadows.default};
  list-style: none;
  margin: 0;
  transition: ${theme.defaultTransition};
  position: absolute;
  width: 100%;
  padding: ${theme.gutter}px ${sidebarGutter};
  margin-top: -${doubleSpaced}px;

  ${({ isOpen }) =>
    !isOpen &&
    css`
      opacity: 0;
      pointer-events: none;
      padding-top: 0;
    `}

  ${breakpoints.tablet`
    box-shadow: 0 0 0;
    position: relative;
    top: unset;
    right: unset;
    left: unset;
    padding-left: 0;
    padding-right: 0;

    ${({ isOpen }) =>
      !isOpen &&
      css`
        opacity: 1;
        pointer-events: auto;
      `}
  `}
`;

const sidebarLi = css`
  padding: ${doubleSpaced}px 0;
  color: ${theme.primary};
  border-bottom: 1px solid ${theme.borders.medium};

  &:last-child {
    border-bottom: 0;
  }

  a {
    color: ${theme.primary};
    font-size: ${theme.fonts.sizes.link}px;
    text-transform: uppercase;
  }
`;

export const StyledSidebar = styled.div`
  position: relative;
`;

export const SidebarTitle = styled(Heading1)`
  border-bottom: 1px solid ${theme.borders.normal};
  position: relative;
  padding: ${theme.gutter}px ${sidebarGutter} ${doubleSpaced}px 0;
  margin: 0 0 ${doubleSpaced}px;
`;

export const SidebarLinks = styled.ul`
  ${sidebarUl}
`;

export const SidebarLink = styled.li`
  ${sidebarLi}
`;

// For prerendered lists
export const SidebarChildren = styled.div`
  ul {
    ${sidebarUl}

    li {
      ${sidebarLi}
    }
  }
`;

export const SidebarIcon = styled(Icon)`
  position: absolute;
  color: ${theme.colors['black-darker']} !important;
  font-size: ${theme.fonts.sizes.mid}px !important;
  right: 3px;
  top: ${theme.gutter + 7}px;
  transition: ${theme.defaultTransition};
`;
