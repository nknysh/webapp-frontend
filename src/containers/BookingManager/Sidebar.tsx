import React from 'react';
import styled from 'styled-components';

export interface ISidebarComponentProps {
  className?: string;
}

const SidebarComponent = (props: ISidebarComponentProps) => {
  return (
    <div className={`booking-manager-sidebar ${props.className}`}>
      <label>Booking</label>
      <ul>
        <li>Summary</li>
        <li>Edit Booking</li>
        <li>Proposals</li>
        <li>Comments</li>
        <li>Terms & Conditions</li>
        <li>Activity Log</li>
      </ul>
    </div>
  );
};

export const Sidebar = styled(SidebarComponent)`
  label {
    font-size: 12px;
  }

  ul {
    list-style: none;
    padding: 0;
    margin: 0;
  }

  ul li {
    padding: 8px;
    background: #ccc;
    margin: 8px 0px;
  }
`;
