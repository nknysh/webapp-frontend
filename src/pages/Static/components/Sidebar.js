// Libraries
import React from 'react';
import { Link } from 'react-router-dom';

// App
import { colors } from 'styles';

// Components
import { Styled } from 'components';

const Container = Styled.View.extend`
  width: 300px;
  margin-right: 150px;
`;

const Title = Styled.H4.extend`
  color: ${colors.gold10};
  margin-bottom: 20px;
`;

const Items = Styled.View.extend`
`;

const Line = Styled.View.extend`
  height: 2px;
  background-color: ${colors.gray15};
`;

const Item = Styled.View.extend`
`;

const Name = Styled.H7.extend`
  margin-vertical: 20px;
`;

const Sidebar = ({ title, items }) => (
  <Container>
    <Title>{title}</Title>
    <Line />
    <Items>
      {items.map(({ name, tag }, index) => (
        <Item key={index}>
          <Link to={`#${tag}`} style={{ marginTop: 20, marginBottom: 20 }}>
            <Name>{name}</Name>
          </Link>
          <Line />
        </Item>
      ))}
    </Items>
  </Container>
);

export default Sidebar;
