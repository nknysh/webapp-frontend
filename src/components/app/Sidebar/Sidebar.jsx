import React, { useState } from 'react';
import { isEmpty } from 'ramda';

import { useCurrentWidth } from 'effects';
import { mapWithIndex } from 'utils';

import { propTypes, defaultProps } from './Sidebar.props';
import { StyledSidebar, SidebarTitle, SidebarLinks, SidebarLink, SidebarChildren, SidebarIcon } from './Sidebar.styles';

export const Sidebar = ({ title, links, isOpen: isOpenProp, children, ...props }) => {
  const { isMobile } = useCurrentWidth();
  const [isOpen, setIsOpen] = useState(isOpenProp);

  const onToggle = () => setIsOpen(!isOpen);
  const onClick = () => setIsOpen(false);

  const hasLinks = !isEmpty(links) || children;

  const renderIcon = () =>
    isMobile && hasLinks && <SidebarIcon>{isOpen ? 'arrow_drop_up' : 'arrow_drop_down'}</SidebarIcon>;

  const renderTitle = () =>
    title && (
      <SidebarTitle onClick={onToggle}>
        {title}
        {renderIcon(isOpen, isMobile)}
      </SidebarTitle>
    );

  const renderLink = (link, i) => (
    <SidebarLink key={i} onClick={onClick}>
      {link}
    </SidebarLink>
  );

  const renderLinks = () => hasLinks && <SidebarLinks isOpen={isOpen}>{mapWithIndex(renderLink, links)}</SidebarLinks>;

  const renderChildren = () =>
    children && (
      <SidebarChildren onClick={onClick} isOpen={isOpen}>
        {children}
      </SidebarChildren>
    );

  return (
    <StyledSidebar {...props}>
      {renderTitle()}
      {renderLinks()}
      {renderChildren()}
    </StyledSidebar>
  );
};

Sidebar.propTypes = propTypes;
Sidebar.defaultProps = defaultProps;

export default Sidebar;
