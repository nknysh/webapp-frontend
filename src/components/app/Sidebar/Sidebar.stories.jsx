import React from 'react';
import { repeat } from 'ramda';

import { storiesOf } from '@storybook/react';
import { withKnobs, text, number } from '@storybook/addon-knobs';

import Sidebar from './Sidebar';

const testLink = <a key={Date.now()}>A Link</a>;

storiesOf('Sidebar', module)
  .addDecorator(withKnobs)
  .add('default', () => {
    return (
      <Sidebar
        links={repeat(
          testLink,
          number('Number of items', 3, {
            range: true,
            min: 0,
            max: 10,
            step: 1,
          })
        )}
        title={text('Sidebar title', 'Sidebar')}
      />
    );
  })
  .add('with children', () => {
    return (
      <Sidebar title={text('Sidebar title', 'Sidebar')}>
        {repeat(
          testLink,
          number('Number of items', 3, {
            range: true,
            min: 0,
            max: 10,
            step: 1,
          })
        )}
      </Sidebar>
    );
  });
