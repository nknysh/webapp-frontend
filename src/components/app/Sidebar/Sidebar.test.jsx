import React from 'react';
import { repeat } from 'ramda';

import Sidebar from './Sidebar';

const getComponent = props => shallow(<Sidebar {...props} />);

const testLink = <a>foo</a>;

describe('<Sidebar />', () => {
  describe('render', () => {
    it('matches snapshots', () => {
      expect(getComponent()).toMatchSnapshot();
      expect(getComponent({ title: 'A Sidebar' })).toMatchSnapshot();
      expect(getComponent({ title: 'A Sidebar', links: repeat(testLink, 5) })).toMatchSnapshot();
      expect(getComponent({ title: 'A Sidebar', children: repeat(testLink, 5) })).toMatchSnapshot();
    });
  });
});
