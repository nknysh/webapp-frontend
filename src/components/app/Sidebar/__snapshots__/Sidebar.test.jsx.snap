// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> render matches snapshots 1`] = `<Sidebarstyles__StyledSidebar />`;

exports[`<Sidebar /> render matches snapshots 2`] = `
<Sidebarstyles__StyledSidebar>
  <Sidebarstyles__SidebarTitle
    onClick={[Function]}
  >
    A Sidebar
  </Sidebarstyles__SidebarTitle>
</Sidebarstyles__StyledSidebar>
`;

exports[`<Sidebar /> render matches snapshots 3`] = `
<Sidebarstyles__StyledSidebar>
  <Sidebarstyles__SidebarTitle
    onClick={[Function]}
  >
    A Sidebar
  </Sidebarstyles__SidebarTitle>
  <Sidebarstyles__SidebarLinks
    isOpen={false}
  >
    <Sidebarstyles__SidebarLink
      key="cb6370c219aa53ec13b2d09141b7982a361f5b6d"
      onClick={[Function]}
    >
      <a>
        foo
      </a>
    </Sidebarstyles__SidebarLink>
    <Sidebarstyles__SidebarLink
      key="cb6370c219aa53ec13b2d09141b7982a361f5b6d"
      onClick={[Function]}
    >
      <a>
        foo
      </a>
    </Sidebarstyles__SidebarLink>
    <Sidebarstyles__SidebarLink
      key="cb6370c219aa53ec13b2d09141b7982a361f5b6d"
      onClick={[Function]}
    >
      <a>
        foo
      </a>
    </Sidebarstyles__SidebarLink>
    <Sidebarstyles__SidebarLink
      key="cb6370c219aa53ec13b2d09141b7982a361f5b6d"
      onClick={[Function]}
    >
      <a>
        foo
      </a>
    </Sidebarstyles__SidebarLink>
    <Sidebarstyles__SidebarLink
      key="cb6370c219aa53ec13b2d09141b7982a361f5b6d"
      onClick={[Function]}
    >
      <a>
        foo
      </a>
    </Sidebarstyles__SidebarLink>
  </Sidebarstyles__SidebarLinks>
</Sidebarstyles__StyledSidebar>
`;

exports[`<Sidebar /> render matches snapshots 4`] = `
<Sidebarstyles__StyledSidebar>
  <Sidebarstyles__SidebarTitle
    onClick={[Function]}
  >
    A Sidebar
  </Sidebarstyles__SidebarTitle>
  <Sidebarstyles__SidebarLinks
    isOpen={false}
  />
  <Sidebarstyles__SidebarChildren
    isOpen={false}
    onClick={[Function]}
  >
    <a>
      foo
    </a>
    <a>
      foo
    </a>
    <a>
      foo
    </a>
    <a>
      foo
    </a>
    <a>
      foo
    </a>
  </Sidebarstyles__SidebarChildren>
</Sidebarstyles__StyledSidebar>
`;
